cunumeric/thunk.py

```diff
@@ -65,120 +70,124 @@ def __numpy_array__(self) -> npt.NDArray[Any]:
         ...
 
     @abstractmethod
-    def imag(self):
+    def imag(self) -> NumPyThunk:
         ...
 
     @abstractmethod
-    def real(self):
+    def real(self) -> NumPyThunk:
         ...
 
     @abstractmethod
-    def conj(self):
+    def conj(self) -> NumPyThunk:
         ...
 
     @abstractmethod
-    def convolve(self, v, out, mode) -> None:
+    def convolve(self, v: Any, out: Any, mode: Any) -> None:
```

Comment:
Essentially any parameter that would otherwise be nice to expect `NumpyThunk` has to be `Any` instead, because methods make assumptions about getting one subclass or another. 

Does `NumpyThunk` need to be an ABC? A protocol might work better here. Additionally we could consider paring `NumpyThunk` down to only encompass interfaces that are truly shared and identical between eager and deferred.

Reply:
> Does NumpyThunk need to be an ABC? A protocol might work better here.

I don't think that's crucial, whatever can stand in for an "interface-only class" would suffice IMHO.